File: Code/PythonCode/1 Two Sum.py
from typing import List


# Solution1 暴力穷举法
class Solution1:
    @staticmethod
    def twoSum(nums: List[int], target: int) -> List[int]:
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                if nums[i] + nums[j] == target:
                    return [i, j]


# Solution2 哈希表HashTable法
class Solution2:
    @staticmethod
    def twoSum(nums: List[int], target: int) -> List[int]:
        hashtable = dict()
        for i, num in enumerate(nums):
            if target - num in hashtable:
                return [hashtable[target - num], i]
            hashtable[nums[i]] = i
        return []


def test_twoSum(solution):
    nums = [2, 7, 11, 15]
    target = 9
    assert solution.twoSum(nums, target) == [0, 1]
    nums = [3, 2, 4]
    target = 6
    assert solution.twoSum(nums, target) == [1, 2]
    nums = [3, 3]
    target = 6
    assert solution.twoSum(nums, target) == [0, 1]
    nums = [3, 2, 3]
    target = 6
    assert solution.twoSum(nums, target) == [0, 2]
    print("通过")


test_twoSum(solution=Solution1)
test_twoSum(solution=Solution2)

File: Code/C++Code/1 Two Sum.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;
//Solution1 暴力穷举法
class Solution1 {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                if (nums[i] + nums[j] == target) {
                    return { i, j };
                }
            }
        }
        return {};
    }
};
//Solution2 哈希表HashTable法
class Solution2 {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int, int> a; //建立hash表a存放数组元素
        vector<int> b(2, -1); //建立数组b存放结果
        for (int i = 0; i < nums.size(); i++)
            a.insert(map<int, int>::value_type(nums[i], i));
            // a[nums[i]] = i;
        for (int i = 0; i < nums.size(); i++)
        {
            if (a.count(target - nums[i]) > 0 && (a[target - nums[i]] != i))
            {
                b[0] = i;
                b[1] = a[target - nums[i]];
                break;
            }          
        }
        return b;
    };
};

int main() {
    // 创建 Solution 类的一个实例
    Solution1 solution1;
    Solution2 solution2;

    // 创建一个整数向量 nums，并初始化它
    vector<int> nums = {3,2,4};

    // 设置目标整数 target
    int target = 6;

    // 调用 twoSum 方法，并传入 nums 和 target 作为参数
    vector<int> hash,result = solution2.twoSum(nums, target);

    // 打印结果
    cout << "[" << result[0] << ", " << result[1] << "]" << endl;

    return 0;
}
